src/model_theory/rings/language.lean

```diff
@@ -0,0 +1,106 @@
+/-
+Copyright (c) 2022 Joseph Hua. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Joseph Hua
+-/
+import model_theory.terms_and_formulas
+
+/-!
+# The language of rings
+
+## Main Definitions
+* A `first_order.language.ring.L` defines the language of rings,
+  which consists of 0,1,-,+,*.
+-/
+universes u v u'
+
+namespace first_order
+namespace language
+namespace ring
+
+/-- The constant symbols in ring.L -/
+inductive consts : Type*
+| zero : consts
+| one : consts
+
+/-- The unary function symbols in ring.L-/
+inductive unaries : Type*
+| neg : unaries
+
+/-- The binary function symbols in ring.L-/
+inductive binaries : Type*
+| add : binaries
+| mul : binaries
+
+/-- All function symbols in ring.langauge-/
+def functions : ℕ → Type*
+| 0 := consts
+| 1 := unaries
+| 2 := binaries
+| (n + 3) := pempty
+
+instance : inhabited consts := ⟨ consts.zero ⟩
+instance : inhabited unaries := ⟨ unaries.neg ⟩
+instance : inhabited binaries := ⟨ binaries.add ⟩
+instance : inhabited (functions 0) := ⟨ consts.zero ⟩
+
+/-- The language of rings -/
+def L : language.{u v} :=
+{ functions := functions,
+  relations := λ n, pempty }
+
+variable {α : Type u'}
+
+@[simp] instance : has_zero (L.term α) := ⟨ @func L _ 0 consts.zero fin_zero_elim ⟩
+
+@[simp] instance : has_one (L.term α) := ⟨ @func L _ 0 consts.one fin_zero_elim ⟩
+
+@[simp] instance : has_neg (L.term α) := ⟨ λ x, @func L _ 1 unaries.neg (λ _, x) ⟩
+
+@[simp] instance : has_add (L.term α) :=
+⟨ λ x y, @func L _ 2 binaries.add (λ i, match i with | ⟨0, _⟩ := x | ⟨n+1, _⟩ := y end) ⟩
+
+@[simp] instance : has_mul (L.term α) :=
+⟨ λ x y, @func L _ 2 binaries.mul (λ i, match i with | ⟨0, _⟩ := x | ⟨n+1, _⟩ := y end) ⟩
+
+instance : has_pow (L.term α) ℕ := ⟨ λ t n, npow_rec n t ⟩
+
+@[simp] lemma pow_zero (t : L.term α) : t ^ 0 = 1 := rfl
+@[simp] lemma pow_succ {n} (t : L.term α) : t ^ (n + 1) = t * t ^ n := rfl
+
+lemma fin_zero_elim_uniq (f : fin 0 → α) : f = fin_zero_elim := subsingleton.elim _ _
```

Comment:
If we need this lemma, can we move it to somewhere under `data.fin`?

Reply:
oh right I was intending to ask about that. I'll delay this PR for now.